ADR/00xx-component-references.md

```diff
@@ -0,0 +1,147 @@
+
+### Build-service
+
+* [build-service] will propagate the digest from one Component to another as a PR – following the declared `ComponentReferences`. This is a bit like renovatebot wrapped in a controller. The [build-service] acts when it sees that a build pipelinerun has completed successfully for a *referent* image.
+  * If this is the first time it has seen this *referent* updated in this way, it will file a new pull request on the *referer* supplying the pullspec and digest of the new referent, using the same branch name as was used for the *referent* PR.
+  * If a PR on that branch already exists, update that branch with an additional commit including the new *referent* digest.
+
+The [build-service] will also update the PR it filed when the PRs that triggered it are merged or updated. It may update the description and/or it may rebase on the main branch and/or it may issue /retest. Need to figure out what we want here.
+
+### “PR dependency”
+
+TODO - describe exactly how the user specifies that one PR depends on another. PaC will have to pass this through to the PipelineRun as an annotation. For now, just assume it is possible and defined.
+
+* When [integration-service] notices a build from a PR (PR#2) that declares it depends on another PR (PR#1), it will do testing but it will also:
+  * Construct the Snapshot for the test using the image from the triggering build pipeline of PR #2 as well as image found on the latest build pipeline associated with PR#1. This lets a user test PR #2 including unmerged content from PR #1.
+  * Post the test results back on PR#1 in addition to the results it would normally post to PR#2.
+  * Post a special followup check result on PR#2 that “fails” saying, “don’t merge this. It still depends on another.”
+* When [integration-service] notices a build from a PR that declares it depends on another PR but that other PR is merged, it will do testing and it will post a followup check result that “succeeds” saying “this is mergeable, all other PRs that it depends on are merged.”
+
+## Applied to Use Cases
+
+Let's apply the architecture to some use cases, and see how it plays out:
+
+### Common Parent Image
+
+Scenario: an application image depends on a common parent image. The user has 1 Application, and 2 Components. One of them is the parent image. The other image is built FROM that image.
+
+* The app image Component declares that it embeds a reference to the parent image Component, by way of a new `ComponentReference` CR.
+* [integration-service] will always skip testing for parent image update builds, will never promote them, or use them to initiate Releases, but it will promote them to the global candidate list.
+* [build-service] will propagate digest references as a PR to the app image Component repo, by analyzing the available `ComponentReference` CRs in the workspace.
```

Comment:
On pull request in common parent image:
1. Pipelines-as-code will build common parent image
2. build-service will detect that ComponentReference exists and creates pull-request in app repo with the updated digest and Do not merge labels
The Pull-request in app repo should never be merged but used for evaluation of base image update. It should be closed when the base image pull request is Closed/Merged.

On push in common parent image:
1. Pipelines-as-code will build common parent image
2. Build-service will create pull-request in app repository
3. This pull-request is processed in standard way -> tested and merged by user.

That's how it could work. I do not get next line where that pull-request "don't merge this" is created by integration service.

Is this expecting that when the pull-request is merged that we re-use the image built by pull-request pipeline? That's not currently happening. Even the images from Pull-requests are scratched build and being deleted in 5 days based on [STONEBLD-1123](https://issues.redhat.com/browse/STONEBLD-1123)

Reply:
I thought that we had the ability to re-use an image from a PR if it was built from the same commit id as a (later) push event to main. This would only work in certain merge strategies and if the PR was already rebased on main.

But - if this is not already the case, we can drop this from the operator design here. It is not essential. I was just building on top of a concept that I thought we already had in place.

We may want to bring that functionality _in_ as part of future changes if it is really absent today; but again, not critical to this operator story. The flow you describe above makes sense to me.